### PR TITLE
cpu/stm32: Clean up / fix periph_eth

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

Reduced RAM by **10,540 Byte**, reduced ROM by 1,600 Byte (in `examples/gnrc_networking` with `BUILD_IN_DOCKER=1`).

- Fixed bug in RX when incoming frame was too large for a single RX buffer
    - This allowed reducing size of RX buffers to safe ~4.5 KiB RAM
- Fix lost RX IRQs (see https://github.com/RIOT-OS/RIOT/issues/13496)
- Implemented zero copy TX to get rid of TX buffers
    - This freed ~6 KiB of RAM
- Mark memory shared with DMA as `volatile`, as its contents are changed by the DMA without knowledge of the compiler
- Drop `__attribute__((packed))` of DMA descriptor, as the descriptor needs to be aligned on word boundaries for the DMA to work.
    - This safes ~300 B ROM :-)
- Integrate `driver/stm32_eth` into `cpu/stm32/periph/eth` to reduce overhead
    - This safes ~1 KiB ROM :-)
- Use `bitarithm.h`'s defines (e.g. `BIT31` instead of `0x80000000`)
- Rename bit defines to also indicate which word the apply to
    - E.g. previously `RX_DESC_RCH` referred to the control word, while `RX_DESC_LS` referred to the status word. The new names `RX_DESC_CTRL_RCH` and `TX_DESC_STAT_LS` make this clear
- Combine read-modify-write cycles to the same control registers where possible
- Add missing waiting for FIFO flushing
- Drop `volatile` of local (stacked) variable in `isr_eth()`

### Testing procedure

- `examples/gnrc_networking` should still work at least as good as previously on a Nucleo-F767ZI
- Issue described in https://github.com/RIOT-OS/RIOT/issues/13496 should no longer occur

### Issues/PRs references

Fixes https://github.com/RIOT-OS/RIOT/issues/13496

### CODE REVIEW

**Comment by maribu** [comment]
I just added the removal of the `__attribute__((packed))`, I hope no one started to review in between.

**Comment by maribu** [comment]
https://github.com/RIOT-OS/RIOT/blob/48fd1aa3e50c8866223ccd9faf759c364e2f4d2c/drivers/stm32_eth/stm32_eth.c#L82-L89

I wonder why this is checking for the RX status during sending. To me, this should be `stm32_eth_get_tx_status_owned()` rather than `stm32_eth_get_rx_status_owned()`. (But only the RX version is provided currently.)

Maybe @crest42 has an explanation for that, or is this indeed just a typo?

**Comment by maribu** [comment]
I'll split the commits apart a bit

**Comment by maribu** [comment]
Sorry for this PR getting longer and longer. I originally wanted to PR just a minor cleanup and fix, but then I wanted also wanted add this little thing, and then that little thing, and then I got going...

This is now ready for review; I won't add new commits or force push anymore without notice.

Some comparison of this PR to master:

- Saved 1,600 B ROM
- Saved 10,540 B RAM (-10,544 B `.bss`, +4B `.data`)
- Practically no difference in performance
    - This PR ended up serving 100 pings slightly faster, but that was so minor it could just be luck

### This PR

```
$ make BUILD_IN_DOCKER=1 RIOT_VERSION=test BOARD=nucleo-f767zi -C examples/gnrc_networking
[...]
  text	   data	    bss	    dec	    hex	filename
  71236	    180	  18224	  89640	  15e28	/data/riotbuild/riotbase/examples/gnrc_networking/bin/nucleo-f767zi/gnrc_networking.elf
```

```
sudo ping -I eth0 -c 100 -i 0.00001 fe80::c13:23ff:fe10:1642
[...]
--- fe80::c13:23ff:fe10:1642 ping statistics ---
100 packets transmitted, 100 received, 0% packet loss, time 61ms
rtt min/avg/max/mdev = 0.350/0.530/0.793/0.097 ms, ipg/ewma 0.616/0.493 ms
```

### `master`


```
$ make BUILD_IN_DOCKER=1 RIOT_VERSION=test BOARD=nucleo-f767zi -C examples/gnrc_networking
[...]
   text	   data	    bss	    dec	    hex	filename
  72836	    176	  28768	 101780	  18d94	/data/riotbuild/riotbase/examples/gnrc_networking/bin/nucleo-f767zi/gnrc_networking.elf
```

```
$ sudo ping -I eth0 -c 100 -i 0.00001 fe80::c13:23ff:fe10:1642
[...]
--- fe80::c13:23ff:fe10:1642 ping statistics ---
100 packets transmitted, 100 received, 0% packet loss, time 63ms
rtt min/avg/max/mdev = 0.400/0.575/0.839/0.066 ms, ipg/ewma 0.641/0.545 ms
```

**Comment by aabadie** [comment]
Do you think this PR could improve the situation described in #13490 and #13496 ?

**Comment by maribu** [comment]
The issue described in https://github.com/RIOT-OS/RIOT/issues/13490 is still present. I think I have an idea how to fix this. I agree that it is quite annoying that the initialization of the driver does not work reliable.

I don't quite understand what the underlying issue in https://github.com/RIOT-OS/RIOT/issues/13496 is, so I don't know if this PR does fix it. My guess would be, that it doesn't. Maybe @kfessel can give this PR a test to check?

**Comment by maribu** [comment]
> I don't quite understand what the underlying issue in #13496 is, so I don't know if this PR does fix it. My guess would be, that it doesn't. Maybe @kfessel can give this PR a test to check?

I think I understand the issue and this PR doesn't solve it. My guess is the issue is the following:

1. While the CPU is halted, the Ethernet DMA is still working and caching incoming pings.
2. The corresponding RX interrupt are lost (except for the last one) while the MCU is stopped
3. After `cont`, each now incoming ping will generate an RX interrupt, but the ISR will only fetch the oldest frames. So the buffer will never get completely drained.

Let me check if I can add a simple fix for this.

**Comment by maribu** [comment]
Handling the lost RX IRQs indeed seems to solve https://github.com/RIOT-OS/RIOT/issues/13496

**Comment by kfessel** [comment]
> Handling the lost RX IRQs indeed seems to solve #13496

i just tested this my result is: partialy fixed (the rest might be hardware) and worsened 

it works for short stops of the processor (in gdb ctrl-c and contiune  within of 2 seconds): fixed no extra delay

it got worse for longer stops ( > 4seconds) in that it entirely stop replying.  This seem to happen faster if I increase the ping size to 128. (ping -6 -s 128 <addr>)

**Comment by crest42** [comment]
Sorry for the late response. The rx/tx stuff should be a typo. Good catch !

**Comment by maribu** [comment]
Handling not only lost, but also spurious IRQs seems to solve the issue completely.

May I squash the last commit right away, so that I can fix the commit message (to also include "spurious")?

**Comment by maribu** [comment]
> May I squash the last commit right away, so that I can fix the commit message (to also include "spurious")?

No one complaint, so I amended the last commit to also handle spurious IRQs.

@kfessel: Please check again.

**Comment by kfessel** [comment]
@maribu: Result: wunderbar
previously failing tests (big (ok 128b is not that big) pings and long pauses) are working just fine.

Big Thanks , great work, (will be a great help when i need the debugger) 

i even could not reproduce the initialization issue #13490 (may also be fixed) - just got it while writing this and playing with the reset button: is not solved but somehow decreased

I get some debug messages: might be leftovers from development?

**Comment by maribu** [comment]
> i even could not reproduce the initialization issue #13490 (may also be fixed)

That would be great. But I fear you just have been lucky. I haven't touched the initialization routine.

File: cpu/stm32/periph/eth.c [hunk 1]
```diff
@@ -14,23 +15,24 @@
...
 #include "iolist.h"
+#include "luid.h"
+#include "mutex.h"
 #include "net/ethernet.h"
-
+#include "net/netdev/eth.h"
 #include "periph/gpio.h"
 
-#define ENABLE_DEBUG (0)
+#define ENABLE_DEBUG (1)
```
Comment (maribu) [inline cpu/stm32/periph/eth.c:33]: TODO: Disable debug again. (Slipped in.)

**Comment by kfessel** [comment]
Travis seems to have found a Typo and something that looks like one 

looks like but is a variable name output type 
>cpu/stm32/include/periph_cpu.h:366: ot  ==> to, of, or

there is something:
>cpu/stm32/include/periph_cpu.h:1092: occured  ==> occurred

and murdock needs a push

**Comment by maribu** [comment]
> looks like but is a variable name output type
> 
> > cpu/stm32/include/periph_cpu.h:366: ot  ==> to, of, or

I agree this is a false positive. The spell checks are only warnings and can be ignored.

> there is something:
> 
> > cpu/stm32/include/periph_cpu.h:1092: occured  ==> occurred

Fixed

> Travis seems to like the last two commits squashed

Since nobody screamed stop, I will do a squash now (and squash the typo fix in the comment right in).

**Comment by maribu** [comment]
Both @kfessel and I have used this PR extensively and it has proven to work pretty reliable, except for the pre-existing and still present initialization bug.

So with this being pretty well tested, only a code review is needed here.

File: cpu/stm32/periph/eth.c [hunk 2]
```diff
@@ -235,126 +234,281 @@ int stm32_eth_init(void)
...
+        if (status & RX_DESC_STAT_LS) {
+            break;
+        }
+        i = i->desc_next;
     }
 
-    /* clear status field */
-    tx_curr->status &= 0x0fffffff;
+    /* bits 16-29 contain the frame length including 4 B frame check sequence */
+    return ((status >> 16) & 0x3fff) - 4;
```
Comment (benpicco) [inline cpu/stm32/periph/eth.c:347]:
```suggestion
    return ((status >> 16) & 0x3fff) - ETHERNET_FCS_LEN;
```

File: cpu/stm32/periph/eth.c [hunk 3]
```diff
@@ -46,40 +48,43 @@
...
+ * be a multiple of 16 Byte. For 64 bit memory interfaces need the size to be
+ * a multiple of 8 Byte, for 32 bit a multiple of 4 byte is sufficient. */
+#warning "ETH_RX_BUFFER_SIZE is not a multiple of 16. (See comment above.)"
+#endif
+
+#if ETH_RX_DESCRIPTOR_COUNT * ETH_RX_BUFFER_SIZE < 1524U
+#warning "Total RX buffers lower than MTU, you won't receive huge frames!"
+#endif
+
+#define MINIMUM(a, b) (((a) <= (b)) ? (a) : (b))
```
Comment (benpicco) [inline cpu/stm32/periph/eth.c:73]: Why not `MIN()`?

Comment (maribu) [reply cpu/stm32/periph/eth.c:73]: I was told that `#define MIN 60` (as in `HOUR`, `MIN`, `SEC`) is also popular and this avoids this ambiguity. But I have no idea why one would use minutes in an Ethernet driver. So if you prefer `MIN` over `MINIMUM`, I could change this.

Comment (benpicco) [reply cpu/stm32/periph/eth.c:73]: `git grep "define MIN("` shows that the `MIN()` macro is actually pretty common.

File: cpu/stm32/periph/eth.c (hunk 2 above)
Comment (maribu) [reply cpu/stm32/periph/eth.c]: Done

File: cpu/stm32/periph/eth.c (hunk 3 above)
Comment (maribu) [reply cpu/stm32/periph/eth.c]: Done

Comment (maribu) [reply cpu/stm32/periph/eth.c]: IMO, we should give the `MIN()` macro in `kernel_defines.h` another spin. I think half of the maintainers blocking that PR back then are no longer active, so chances are good a PR would now make it.

**Comment by maribu** [comment]
@benpicco: Thanks for reviewing :-)
@kfessel: Thanks for testing :-)

**Comment by JannesVolkens** [comment]
This PR somehow causes LwIP to always send two identical messages instead of one when using `sock_udp_send` and IPv4. :/

**Comment by maribu** [comment]
> This PR somehow causes LwIP to always send two identical messages instead of one when using `sock_udp_send` and IPv4. :/

Could you create a minimum application to ease with reproducing the issue? Using the `txtsnd` command to send raw Ethernet frames to a Linux host who is capturing frames with Wireshark I only get a single frame per each transmission. Same with `ping6` on top of GNRC: One ICMPv6 message per ping sent. Also the layer two network statistics collected via module `netstats_l2` and shown via the `ifconfig` command look sensible: No errors, number of transmission and total amount of bytes sense are as expected.

From all that I can tell it looks like the driver behaves as it should.

**Comment by JannesVolkens** [comment]
> Could you create a minimum application to ease with reproducing the issue? Using the txtsnd command to send raw Ethernet frames to a Linux host who is capturing frames with Wireshark I only get a single frame per each transmission. Same with ping6 on top of GNRC: One ICMPv6 message per ping sent. Also the layer two network statistics collected via module netstats_l2 and shown via the ifconfig command look sensible: No errors, number of transmission and total amount of bytes sense are as expected.

I have also checked this and everything is sent correctly here.

I connected two RIOT nodes and ran the test under `tests/lwip` for IPv4. To give the nodes an IPv4 address I added 
```
struct netif *iface = netif_list;
IP4_ADDR(&iface->ip_addr, 192, 168, 1, 1X);
```
to the `main`, because #14688 is not yet merged.

Now when I send a packet the receiver node gets two packets:

sender:
```
2020-08-27 09:19:14,495 #  udp send 192.168.1.11:44 abcdef
2020-08-27 09:19:14,499 # Success: send 3 byte over UDP to 192.168.1.11:44
```

receiver:
```
2020-08-27 09:19:14,498 # Received UDP data from [192.168.1.12]:45
2020-08-27 09:19:14,500 # 00000000  AB  CD  EF
2020-08-27 09:19:14,504 # Received UDP data from [192.168.1.12]:45
2020-08-27 09:19:14,506 # 00000000  AB  CD  EF
```

I also checked with wireshark and captured two packets for each send.

**Comment by maribu** [comment]
I can reproduce the issue and it is indeed the driver sending twice. The bug is triggered when only a single element is in the `iolist` passed for sending. It seems that there is a race condition in when the DMA clears the own bit and when it reads the own bit of the next DMA descriptor. I used the TER bit to tell the DMA after the last DMA descriptor to start looking again at the first descriptor, so that the driver does not need to keep track of the currently used descriptor. If there is only one chunk in the `iolist`, the DMA directly goes back to the same descriptor and finds the own bit still set due to this hardware race condition.

I opened a pull request reverting the behavior to not using the TER bit and instead storing a pointer to the currently used DMA descriptor. With that, TX works again as expected.

**Pull request merged**